tests/ut_libdsme.c

```diff
@@ -26,210 +27,425 @@
  * some glibc versions seems to mistakenly define ucred behind __USE_GNU;
  * work around by #defining _GNU_SOURCE
  */
-#ifndef __cplusplus
 #define _GNU_SOURCE
-#endif
-
-#include <check.h>
-#include <fcntl.h>
-#include <stdlib.h>
-#include <stdio.h>
-#include <unistd.h>
-#include <sys/sem.h>
-#include <sys/stat.h>
-#include <sys/wait.h>
 
 #include "../include/dsme/messages.h"
 #include "../include/dsme/protocol.h"
 #include "../include/dsme/state.h"
 
-static const char* const TMP_SOCKFILE = "/tmp/ut_libdsme.sock";
+#include <sys/stat.h>
+#include <sys/wait.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <unistd.h>
+#include <errno.h>
+#include <poll.h>
+#include <syslog.h>
+#include <ctype.h>
+#include <getopt.h>
+
+#include <check.h>
+
+/* ------------------------------------------------------------------------- *
+ * Diagnostic Logging
+ * ------------------------------------------------------------------------- */
+
+static int log_level = LOG_WARNING;
+
+static const char * const log_tag[] = {
+    [LOG_EMERG]   = "X: ",
+    [LOG_ALERT]   = "A: ",
+    [LOG_CRIT]    = "C: ",
+    [LOG_ERR]     = "E: ",
+    [LOG_WARNING] = "W: ",
+    [LOG_NOTICE]  = "N: ",
+    [LOG_INFO]    = "I: ",
+    [LOG_DEBUG]   = "D: ",
+};
+
+static void log_emit_(const char *file, int line, const char *func,
+                      int level, const char *fmt, ...)
+{
+    int saved = errno;
+    char *msg = NULL;
+    va_list va;
+    va_start(va, fmt);
+    if( vasprintf(&msg, fmt, va) < 0 )
+        msg = NULL;
+    va_end(va);
+    fprintf(stderr, "%s:%d: %s: %s%s\n",
+            file, line, func,
+            log_tag[level], msg ?: fmt);
```

Comment:
Got wondering what ?: is defined to do but apparently a gnu extension. Could just write it out as "? msg :" to ease readability?

Reply:
The whole point of ?: is that it increases readability by having just one thing once instead of two things that one then needs to mentally verify to be the same thing.

Reply:
I could buy that argument if it's widely used and supported but here it seemed the only instance on the whole repository. Spent myself some time to figure out if it's a valid construct.

But quite minor detail anyhow.

Reply:
There is always the 1st use of whatever construct in one particular git repo even when it is something already commonly used elsewhere. In case of "?:", a quick check from some git repos that I happen to have checkouts yields:

cfiles github.com/sailfishos|xargs grep '? *:'|wc -l
625